R/decostand.R

```diff
@@ -156,61 +156,96 @@
-    if (any(x <= 0, na.rm = na.rm)) {
+    # Error with negative values (note: at this step we always use na.rm=TRUE)
+    if (any(x <= 0, na.rm = TRUE)) {
         stop("'clr' cannot be used with non-positive data: use pseudocount > ",
-             -min(x, na.rm = na.rm) + pseudocount, call. = FALSE)
+             -min(x, na.rm = TRUE) + pseudocount, call. = FALSE)
     }
 
     # In every sample, calculate the log of individual entries.
     # Then calculate
     # the sample-specific mean value and subtract every entries'
     # value with that.
     clog <- log(x)
-    means <- rowMeans(clog, na.rm = na.rm)
+
+    # Calculate sample-wise log means (note: here we always set na.rm=TRUE !)
+    means <- rowMeans(clog, na.rm = TRUE)
+
+    # CLR transformation
     clog <- clog - means
+
+    # Replace missing values with 0
+    if (na.rm && any(is.na(clog))) {
+        cat("Replacing missing values with zero for clr. You can disable this with na.rm=FALSE.")    
+        clog[is.na(clog)] <- 0
+    } 
+    
     attr(clog, "parameters") <- list("means" = means,
                                      "pseudocount" = pseudocount)
     clog
 }
 
 # Modified from the original version in mia R package
 .calc_rclr <-
-    function(x, na.rm = TRUE)
+    function(x, na.rm, ROPT=3, NITER=5, TOL=1e-5, verbose=FALSE, impute=TRUE, ...)
 {
     # Error with negative values
-    if (any(x < 0, na.rm = na.rm)) {
+    if (any(x < 0, na.rm=na.rm)) {
         stop("'rclr' cannot be used with negative data", call. = FALSE)
     }
+
    # Log transform
    clog <- log(x)
+   
    # Convert zeros to NAs in rclr
    clog[is.infinite(clog)] <- NA
+
    # Calculate log of geometric mean for every sample, ignoring the NAs
-   mean_clog <- rowMeans(clog, na.rm = na.rm)
-   # Divide all values by their sample-wide geometric means
-   # Log and transpose back to original shape
-   xx <- log(x) - mean_clog
-   # If there were zeros, there are infinite values after logarithmic transform.
-   # Convert those to zero.
-   xx[is.infinite(xx)] <- 0
-   attr(xx, "parameters") <- list("means" = mean_clog)
-   xx
-}
+   # Always na.rm=TRUE at this step!   
+   means <- rowMeans(clog, na.rm = TRUE)
 
+   ## Divide (or in log-space, reduce) all values by their sample-wide
+   ## geometric means
+   xx <- clog - means
+   attr(xx, "parameters") <- list("means" = means)
+
+   # Impute NAs if impute=TRUE
+   # Otherwise return the transformation with NAs
+   if (impute && any(is.na(xx))){
+   
+     opt_res <- OptSpace(xx, ROPT=ROPT, NITER=NITER, TOL=TOL, verbose=verbose)
+     
+     # recenter the data (the means of rclr can get thrown off since we work on only missing)
+     M_I <- opt_res$X %*% opt_res$S %*% t(opt_res$Y)
+
+     # Center cols to 0
+     M_I <- as.matrix(scale(M_I, center=TRUE, scale=FALSE))
+
+     # Center rows to 0
+     M_I <- as.matrix(t(scale(t(M_I), center=TRUE, scale=FALSE)))
+
+     # Imputed matrix
+     xx <- M_I
+   }
+  
+   return(xx)
```

Comment:
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.